### PR TITLE
DarwinCompatibility: Restore tests to compiling and working

### PR DESCRIPTION
**Pull request by spevans**

### CODE REVIEW

**Comment by spevans** [comment]
@swift-ci test

**Comment by spevans** [comment]
@swift-ci test linux

**Comment by spevans** [comment]
@swift-ci test

**Comment by spevans** [comment]
@swift-ci test

**Comment by spevans** [comment]
@swift-ci test linux

File: TestFoundation/TestMeasurement.swift [hunk 1]
```diff
@@ -8,7 +8,7 @@
 //
 
 class CustomUnit: Unit {
-    required init(symbol: String) {
+    required override init(symbol: String) {
```
Comment (millenomi) [inline TestFoundation/TestMeasurement.swift]: We need a bug for this.

File: TestFoundation/TestNSSet.swift [hunk 2]
```diff
@@ -237,9 +237,14 @@ class TestNSSet : XCTestCase {
         let set: NSSet = NSSet(array: [array, dictionary, innerSet, 1111, 2222, 3333])
         
         let description = NSString(string: set.description)
-        
+
+#if DARWIN_COMPATIBILITY_TESTS
+        XCTAssertTrue((description.substring(to: 2) as NSString).isEqual(to: "{("))
```
Comment (millenomi) [inline TestFoundation/TestNSSet.swift]: These indicate that we have an incorrect signature  or missing overload for `.substring(…)` in SCF. We should fix that instead.

File: TestFoundation/TestNSString.swift [hunk 3]
```diff
@@ -1056,7 +1056,13 @@ class TestNSString: LoopbackServerTest {
         do {
             let path: NSString =  "~/foo/bar/"
             let result = path.standardizingPath
+#if DARWIN_COMPATIBILITY_TESTS
+            // FIXME: Darwin gives unavailable for 'appendingPathComponent'
```
Comment (millenomi) [inline TestFoundation/TestNSString.swift]: This needs to be investigated as well.

File: TestFoundation/Utilities.swift [hunk 4]
```diff
@@ -203,7 +224,11 @@ extension Fixture where ValueType: NSObject & NSCoding {
         let original = try make()
         
         let coder = NSKeyedArchiver(forWritingWith: NSMutableData())
+#if !DARWIN_COMPATIBILITY_TESTS
+        // FIXME: On Darwin, this is defined as @property (readwrite) NSDecodingFailurePolicy decodingFailurePolicy
```
Comment (millenomi) [inline TestFoundation/Utilities.swift]: Something is very wrong here. The current definition of this property on master is:

```
open var decodingFailurePolicy: NSCoder.DecodingFailurePolicy = .raiseException
```

Comment (spevans) [reply TestFoundation/Utilities.swift]: This is when running against native Foundation. On Xcode 10.2.1 the definition comes from :

```
/Applications/Xcode.app//Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/Foundation.framework/Versions/C/Headers/NSKeyedArchiver.h

@property (readwrite) NSDecodingFailurePolicy decodingFailurePolicy API_AVAILABLE(macos(10.11), ios(9.0), watchos(2.0), tvos(9.0));
```

File: TestFoundation/TestMeasurement.swift (hunk 1 above)
Comment (spevans) [reply TestFoundation/TestMeasurement.swift]: I think the correct fix here is to remove the `required` on `Unit.init(symbol: String)` to match Darwin:
https://developer.apple.com/documentation/foundation/unit/1690760-init
Then this `override` is required in both cases.

However removing the `required` leads to this line
`unit = UnitType(symbol: symbol)` 
in `Measurement.init(from decoder: Decoder)` giving the error:

`Constructing an object of class type 'UnitType' with a metatype value must use a 'required' initializer`

File: TestFoundation/Utilities.swift (hunk 4 above)
Comment (spevans) [reply TestFoundation/Utilities.swift]: I think I've fixed this correctly in https://github.com/apple/swift-corelibs-foundation/pull/2324

**Comment by spevans** [comment]
@swift-ci test

File: TestFoundation/TestMeasurement.swift (hunk 1 above)
Comment (spevans) [reply TestFoundation/TestMeasurement.swift]: Filed https://bugs.swift.org/browse/SR-10904

**Comment by spevans** [comment]
@swift-ci test

**Comment by spevans** [comment]
@swift-ci test

**Comment by spevans** [comment]
@swift-ci test

File: TestFoundation/TestNSString.swift (hunk 3 above)
Comment (spevans) [reply TestFoundation/TestNSString.swift]: Filed https://bugs.swift.org/browse/SR-10916

File: TestFoundation/TestNSSet.swift (hunk 2 above)
Comment (spevans) [reply TestFoundation/TestNSSet.swift]: This is also https://bugs.swift.org/browse/SR-10916 which I have rewritten to use `.hasPrefix()` and `.hasSuffix()`

**Comment by spevans** [comment]
@swift-ci test

**Comment by spevans** [comment]
@swift-ci test linux

**Comment by spevans** [comment]
@millenomi All of the workarounds have bugs filed for them now so I think this is ok to go in.

**Comment by spevans** [comment]
@swift-ci test linux

**Comment by spevans** [comment]
@swift-ci test linux

**Pull request merged**